### PR TITLE
Optionally add left icon in the KeyValuePage's titlebar

### PR DESCRIPTION
**Pull request by roygbyte**
I've wanted this for a while :)

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9495)

### CODE REVIEW

File: frontend/ui/widget/keyvaluepage.lua [hunk 1]
```diff
@@ -14,6 +14,10 @@ Example:
             {"Press me", "will invoke the callback",
              callback = function() print("hello") end },
         },
+         title_bar_left_icon = "appbar.menu",
+         title_bar_left_icon_tap_callback = function()
+            -- Do something on left icon callbac
```
Comment (NiLuJe) [inline frontend/ui/widget/keyvaluepage.lua:19]: nit: callback

File: frontend/ui/widget/keyvaluepage.lua [hunk 2]
```diff
@@ -439,6 +443,12 @@ function KeyValuePage:init()
         bottom_line_color = Blitbuffer.COLOR_DARK_GRAY,
         bottom_line_h_padding = padding,
         close_callback = function() self:onClose() end,
+            left_icon = self.title_bar_left_icon,
```
Comment (NiLuJe) [inline frontend/ui/widget/keyvaluepage.lua:446]: nit: extra indent

**Comment by hius07** [comment]
In the Menu widget we have it in a little bit different way:
https://github.com/koreader/koreader/blob/96930f230f1a6a3ba51e078a7d344428d3fd2707/frontend/ui/widget/menu.lua#L678-L681

**Comment by zwim** [comment]
What is the use case for this? If there is no real use case, this would be dead code. Do we want to add dead code?

But anyway. If there is real need, maybe a hold calback might be of interest, too?

Have you seen my userpatch https://gist.github.com/zwim/1edcd34ef8a59166f203d5ee8c08f7e3 ?

**Comment by roygbyte** [comment]
Use case is in Newsdownloader replacement I'm working on. 

Had not seen your user patch. Do you suggest using the same naming conventions, as in the patch, for this change?

**Comment by poire-z** [comment]
That's some generic top left button (to be used or not) for the KeyValuePage widget, just like the menu has it - so, dead or not used yet, no problem for me adding it.

**Comment by roygbyte** [comment]
I would like to add that, broadly speaking, it is useful to have buttons available for operating on a list of values. Without this patch, there's no way to operate on the list of values, unless through some gesture or by inserting a value at the beginning or end of a list with some special function (as I have previously done). Use-case operations could include:

- Deleting the list
- Refreshing the list
- Adding to the list
- Filtering the list

Anyways--maybe there's no immediate use for this function, but to me this widget is incomplete without a button.

**Comment by zwim** [comment]
I really have nothing against adding functionality. But it seems a bit odd to add code, which is not used anywhere (especially as you have some ideas).

Maybe you might toggle the PR to draft, _until_ the rest of your work is ready. I can imagine your intention to break the big work into small pieces, to check, test, see how these will be accepted, and adapt them. This would also be the case with a draft PR.

When you are ready, all the corresponding PRs can be turned to "ready for review". This way might give more insight of what is happening and when in the code base.

**Comment by Frenzie** [comment]
Some code makes sense to add regardless if it's used yet, but a screenshot would be nice. ;-)

**Comment by roygbyte** [comment]
I didn't know PR's could be in a draft stage. That's great--thanks for the advice.

Below for screenshots: 

Screenshot from 2022-09-09 20-32-19

Screenshot from 2022-09-09 20-32-37

I'd rather use a "+" for the icon in this scenario, but haven't found that one yet.

**Comment by hius07** [comment]
https://github.com/koreader/koreader/blob/d972b7fcfaac8b83c53cf7826ce74e1644a96f32/frontend/apps/cloudstorage/cloudstorage.lua#L58

**Comment by poire-z** [comment]
@hius07 : would this be how you would have implemented it ?
If not, don't you miss a left icon (ie. for refreshing current listing) in OPDS or Cloud storage ? :) so you could quickly implement it and we can close this PR :)

**Comment by hius07** [comment]
> don't you miss a left icon (ie. for refreshing current listing) in OPDS or Cloud storage ?

They are Menu widgets, not KVP.

In the Menu we do it this way:
https://github.com/koreader/koreader/blob/724db8775561651abbf24650b95b2d03a07d961d/frontend/ui/widget/menu.lua#L677-L679
https://github.com/koreader/koreader/blob/724db8775561651abbf24650b95b2d03a07d961d/frontend/ui/widget/menu.lua#L1388-L1392
Are there any preferences comparing with this PR?

**Comment by NiLuJe** [comment]
I'd unify behavior/style with Menu, yeah?

**Comment by roygbyte** [comment]
> I'd unify behavior/style with Menu, yeah?

I'm ultimately fine with this. But I would like to observe that the KVP layout has another left icon: the left arrow icon. So having "title" in the attribute does help distinguish between the two, plus it provides some added context (indicating the title bar icon is what's being set).

**Comment by roygbyte** [comment]
Superseded by #9713

**Pull request closed without merging**